base/arrayshow.jl

```diff
@@ -0,0 +1,421 @@
+                    end
+                end
+                if last(cr) == last(indc)
+                    i < last(indr) && print(io, "; ")
+                elseif cdots
+                    print(io, " \u2026 ")
+                end
+            end
+        end
+        last(rr) != nr && rdots && print(io, "\u2026 ; ")
+    end
+    print(io, "]")
+end
+
+
+_show(io::IO, X::AbstractArray, prefix::String) =
+    show_nd(io, X, (io, slice) -> _show(io, slice, prefix), false)
+
+# a specific call path is used to show vectors (show_vector)
+_show(::IO, ::AbstractVector, ::String) = error("_show(::IO, ::AbstractVector, ::String) is not implemented")
+
+_show(io::IO, X::AbstractArray{T,0} where T, prefix::String) = print_array(io, X)
+
+# NOTE: it's not clear how this method could use the :typeinfo attribute
+_show_empty(io::IO, X::Array{T}) where {T} = print(io, "Array{$T}(", join(size(X),','), ')')
+_show_empty(io, X) = nothing # by default, we don't know this constructor
+
+# typeinfo aware (necessarily)
+function show(io::IO, X::AbstractArray)
+    @assert ndims(X) != 1
+    prefix = typeinfo_prefix(io, X)
+    io = IOContext(io, :typeinfo => eltype(X), :compact => true)
+    isempty(X) ?
+        _show_empty(io, X) :
+        _show(io, X, prefix)
+end
+
+### Vector arrays
+
+# typeinfo aware
+# NOTE: v is not constrained to be a vector, as this function can work with iterables
+# in general (it's used e.g. by show(::IO, ::Set))
+function show_vector(io::IO, v, opn='[', cls=']')
+    print(io, typeinfo_prefix(io, v))
+    # directly or indirectly, the context now knows about eltype(v)
+    io = IOContext(io, :typeinfo => eltype(v), :compact => true)
+    limited = get(io, :limit, false)
+    if limited && _length(v) > 20
+        inds = indices1(v)
+        show_delim_array(io, v, opn, ",", "", false, inds[1], inds[1]+9)
+        print(io, "  \u2026  ") # i.e. "  …  "
+        show_delim_array(io, v, "", ",", cls, false, inds[end-9], inds[end])
+    else
+        show_delim_array(io, v, opn, ",", cls, false)
+    end
+end
+
+show(io::IO, X::AbstractVector) = show_vector(io, X)
+
+
+## Logic for displaying type information
+
+# X not constrained, can be any iterable (cf. show_vector)
+function typeinfo_prefix(io::IO, X)
+    typeinfo = typeinfo_context(io)
+    @assert X isa typeinfo
+    # what the context already knows about the eltype of X:
+    # (could use a more specific tool than eltype, maybe with _isleaftype)
+    eltype_ctx = eltype(typeinfo)
+    eltype_X = eltype(X)
+    # Types hard-coded here are those which are created by default for a given syntax
+    if eltype_X == eltype_ctx || !isempty(X) && eltype_X in (Float64, Int, Char, String)
+        ""
+    elseif print_without_params(eltype_X)
+        string(unwrap_unionall(eltype_X).name) # Print "Array" rather than "Array{T,N}"
+    else
+        string(eltype_X)
+    end
+end
+
+typeinfo_context(io::IO) = get(io, :typeinfo, Any)::Union{DataType,UnionAll}
```

Comment:
Is this wrapper really useful? It makes it actually harder to read the code for people who don't know that it's just a call to `get`.

Reply:
I'm fine with removing it, it's mainly a leftover from my successive re-factorization of the implementation of the new feature. But one thing which I will was addressed is that the default value of attributes (`Any` in this case) was set once and for all, instead of repeated at each call site. I have been thinking of defining `getindex(io::IO, ::Symbol)` which would take care of that, but it's not clear that it would be the best way.